### PR TITLE
Task 2: Schedule DaemonSet Pods by default scheduler.

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
part of #59194
https://github.com/kubernetes/features/issues/548

**Release note**:

```release-note
When ScheduleDaemonSetPods is enabled, the DaemonSet controller will delegate Pods scheduling to default scheduler.
```

### CODE REVIEW

**Comment by k82cn** [comment]
@bsalamat @kow3ns , would you help to review whether this solution is OK: in this PR, I just 'hack' `nodeShouldRunDaemonPod` to return 'correct' value, e.g. wantToRun==shouldSchedule==true. And I'd like to simply it to one return value when graduating to beta.

File: pkg/controller/daemon/daemon_controller.go
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: s/enable/enabled/

File: pkg/controller/daemon/daemon_controller.go
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: I would suggest renaming to `nodeSelectionPredicates`.

File: pkg/controller/daemon/daemon_controller.go
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: Please change to:
nodeSelectionPredicates runs a set of predicates that select candidate nodes for the DaemonSet.

File: pkg/controller/daemon/daemon_controller.go
Comment (bsalamat) [inline pkg/controller/daemon/daemon_controller.go]: We don't need to run PodFitsHostPorts and actually your function does not run it. Please remove it from the comment.

**Review by bsalamat — COMMENTED** [review]
Thanks, @k82cn!
Your change looks correct to me and given that we want to keep the old behavior and enable the new behavior with a flag, your change is OK, but in the future when we remove the scheduling logic of the DS controller, we can refactor and simplify the logic.

**Comment by janetkuo** [comment]
Please add a release note.

File: pkg/controller/daemon/daemon_controller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: Please combine this and #59798 into one PR (ok to be separate commits)

File: pkg/controller/daemon/daemon_controller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: So we can't remove all predicates in DS controller even when switching to default scheduler?

**Review by janetkuo — COMMENTED** [review]
~Is #60163 planned to be solved in this PR?~ Never mind I saw you opened #60386

File: pkg/controller/daemon/daemon_controller.go
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: sure; let me combine them :)

File: pkg/controller/daemon/daemon_controller.go
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: As daemonset can use `NodeAffinity` and `NodeSelector` to define a subset of cluster to stat the Daemon Pods, we re-use those predicates.

The advantage is that we can choose which predicate we'd like to re-use according to DaemonSet's requirement instead of keep it align with default scheduler when a new feature introduced.

**Comment by yastij** [comment]
@k82cn @bsalamat - is this one making it for 1.10 ?

**Comment by k82cn** [comment]
It should be in 1.10 :)

**Comment by k82cn** [comment]
@janetkuo , comments addressed, PTAL; btw, I think this also should be in 1.10, would you help to /approve-for-milestone ?

**Comment by janetkuo** [comment]
> I think this also should be in 1.10, would you help to /approve-for-milestone ?

This feature needs to be accepted by 1.10 release team before I approve it for milestone, please see my comment: https://github.com/kubernetes/features/issues/548#issuecomment-370003540

**Comment by kargakis** [comment]
@kubernetes/sig-apps-pr-reviews 
@kubernetes/sig-scheduling-pr-reviews

**Comment by k82cn** [comment]
exception request sent :)

File: pkg/controller/daemon/util/daemonset_util.go
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: This tells the DaemonSet controller to ignore all hostname related RequiredDuringSchedulingIgnoredDuringExecution node affinity specified in its template. 

DaemonSet controller shouldn't do that. When it schedules a pod, it takes node affinity into account, instead of forcing all its pod to be scheduled on each node (by setting `nodeName`).

File: pkg/controller/daemon/util/daemonset_util.go
Comment (janetkuo) [inline pkg/controller/daemon/util/daemonset_util.go]: How does the default scheduler handle affinity conflicts? For example, when the node affinity is specified like this:

1. Schedule on node with label `hostname in [foo]`, and 
1. Schedule on node with label `hostname in [bar]`

Or this case:

1. Schedule on node with label `hostname in [foo]`, and
1. Schedule on node with label `hostname not in [foo]`

**Review by janetkuo — CHANGES_REQUESTED** [review]

**Comment by janetkuo** [comment]
This needs a release note.

**Comment by jberkus** [comment]
@janetkuo @k82cn if the exception is granted, please remember to add the milestone, priority/critical-urgent, and approved-for-milestone to this issue. thanks!

File: pkg/controller/daemon/util/daemonset_util.go
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: In predicates, we use NodeAffinity to identify the nodes we're going to run DaemonSet. So the hostname in this function is one of NodeAffinity.

For your cases, the `nodeSelectionPredicates` will return empty, so it'll not happen :)

File: pkg/controller/daemon/util/daemonset_util.go
Comment (janetkuo) [reply pkg/controller/daemon/util/daemonset_util.go]: Synced offline, the input `nodename` shouldn't have conflict with `affinity` because `nodeSelectionPredicates` is invoked first

File: pkg/controller/daemon/daemon_controller.go
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: As no behavior changed, I'd like to run DaemonSet e2e test with this feature enabled. So logged https://github.com/kubernetes/test-infra/issues/7149 with test-infra team to suggest to run it in a separate CI environment.

File: pkg/controller/daemon/daemon_controller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemon_controller.go]: This predicate checks node name, node selector, node affinity specified in the pod spec. However, it's not obvious from the comment or function name. Would you make it more obvious?

File: pkg/controller/daemon/daemon_controller.go
Comment (janetkuo) [reply pkg/controller/daemon/daemon_controller.go]: Also need to take care of the case when the user specifies `.nodeName` in the pod spec.

File: pkg/controller/daemon/util/daemonset_util.go
Comment (janetkuo) [reply pkg/controller/daemon/util/daemonset_util.go]: Maybe it should be named `Overwrite` or `Replace` instead of `Append`?

Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: done

File: pkg/controller/daemon/util/daemonset_util.go
Comment (k82cn) [reply pkg/controller/daemon/util/daemonset_util.go]: wow, removed empty slice.

File: pkg/controller/daemon/daemon_controller.go
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: done.

File: pkg/controller/daemon/daemon_controller.go
Comment (k82cn) [reply pkg/controller/daemon/daemon_controller.go]: Added test :)

> Also need to take care of the case when the user specifies .nodeName in the pod spec.

As `pod.Spec.NodeName` is not slice, so not mismatch. And suppose user will not update it manually. NodeName is handled in `PodFitsHost`, part of `nodeSelectionPredicates` :)

**Review by janetkuo — APPROVED** [review]
Only some nits

**Comment by janetkuo** [comment]
Is the feature name `NoDaemonSetScheduler` finalized? It looks like DaemonSet won't have any scheduler. 
It's probably better to make it more clear that DaemonSet's pods is scheduled by the default scheduler, such as `ScheduleDaemonSetPods` or `DaemonSetThroughScheduler`.

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
/assign @smarterclayton 
Would you approve the change? Thanks!

File: pkg/controller/daemon/daemon_controller.go
Comment (mfojtik) [inline pkg/controller/daemon/daemon_controller.go]: will this deal with existing daemonset pods during migration? or the daemonset have to be recreated/rescaled?

**Comment by smarterclayton** [comment]
Feature name approved

/approve

File: pkg/controller/daemon/daemon_controller.go
Comment (janetkuo) [reply pkg/controller/daemon/daemon_controller.go]: This only affects Daemon pod creation. Existing pods won't be affected. Existing DaemonSet might be affected when, for example, their pods are killed/dead or new nodes are added.

**Comment by k82cn** [comment]
and ping for milestone approve :)

**Pull request merged**
